File: LPtHW/ex44/ex44e.py
class Other(object):

    def override(self):
        print "OTHER override()"

    def implicit(self):
        print "OTHER implicit()"

    def altered(self):
        print "OTHER altered()"

class Child(object):

    def __init__(self):
        self.other = Other()

    def implicit(self):
        self.other.implicit()

    def override(self):
        print "CHILD override()"

    def altered(self):
        print "CHILD BEFORE OTHER altered()"
        self.other.altered()
        print "CHILD AFTER OTHER altered()"

son = Child()

son.implicit()
son.override()
son.altered()

# Composition allows the same functionality as inheritance, 
# but without the danger of creating dependencies with inheritance.
# This allows for more modular code that is easier to maintain.
# In fact, you can even abstract away Other() into its own module
# and import it!

# instead of parent is-a child, child has-a other.
# Basically we are bringing in one object to serve the other rather than
# have that other interit from the object serving it.

File: Learn_Ruby_the_Hard_Way/ex44/ex44e.rb
class Hammer

  def override
    puts "OTHER override()"
  end

  def implicit
    puts "OTHER implicit()"
  end

  def altered
    puts "OTHER altered()"
  end
end

class Child

  def initialize
    @hammer = Hammer.new
  end

  def implicit
    @hammer.implicit
  end

  def override
    puts "CHILD override()"
  end

  def altered
    puts "CHILD, BEFORE OTHER altered()"
    @hammer.altered
    puts "CHILD, AFTER OTHER altered()"
  end
end

maxwell = Child.new

maxwell.implicit
maxwell.override
maxwell.altered
